### ISSUE TITLE
failed to run script: [string "android.lua"] error extracting assets

### ISSUE THREAD

**Issue by ildar** [koreader/android-luajit-launcher#265]
Built apk but it fails:
```
11-17 10:13:08.234 15867 15893 V luajit-launcher: [NativeThread]: activity gained focus.
11-17 10:13:08.235 15867 15893 V luajit-launcher: [NativeThread]: launching LuaJIT assets
11-17 10:13:08.256 15867 15867 D luajit-launcher: [NativeGlue]: NativeWindowRedrawNeeded: 0xddd2ac60 -- 0xdddbf008
11-17 10:13:08.268 15867 15893 D luajit-launcher-debug: checking write storage permission
11-17 10:13:08.273 15867 15893 I luajit-launcher-debug: Zip file not found, trying raw assets...
11-17 10:13:08.276 15867 15893 E luajit-launcher: [NativeThread]: failed to run script: [string "android.lua"]:2421: error extracting assets
11-17 10:13:08.276 15867 15893 E luajit-launcher: [NativeThread]: Stopping due to previous errors
```

BTW, I can't find any chat, is there one? I joined #luajit IRC channel but it's silent.

**Comment by Frenzie** [comment]
I don't think there's any failure there. See https://github.com/koreader/android-luajit-launcher/tree/b23c380a8eecee6c1f15ab35d20053d12986e51a/examples/helloWorld for an example, and what it says in the README:

> Have a look at KOReader's llapp_main.lua file. You can use it as a starting point for your own app.
>
> The real starting point, called from JNI/C, is the run() function in android.lua. It sets up a few things, namely FFI definitions for the Android native API (since it uses that itself for a few things) and some wrapper functions for logging. Also, it registers the "android" module in package.loaded, so you can access it in your own code via require("android"). It also registers a new package loader which can load Lua code from the activity's asset store, so you can use require() for Lua code stored there.

For chat there's https://gitter.im/koreader/koreader

**Comment by pazos** [comment]
@ildar: this project is just glue code for KOReader. It works on top of NativeActivity framework, which was built to make it easier to port games to android without touching Java and without writting any JNI code.

NativeActivity expects the program to write to its own window and to handle its own input queue. These things are enough to port a game without too much trouble but won't let you access to android frameworks.

This program is useless unless you want to port a luajit program that has its own drawing routines.

If your purpose is to develop an android program that's able to execute lua code you can switch to android SDK and use JNI to setup lua states or use one of the very good available libraries, like https://github.com/alibaba/LuaViewSDK

TL;DR: This repo has no purpose outside KOReader. Of course is still useful to understand NativeActivity, android-native-app-glue and contains a pretty damn good LuaJIT JNI interface, so it could have educational purposes too.

**Comment by Frenzie** [comment]
Also see https://github.com/love2d/love-android for some interesting stuff.

**Comment by pazos** [comment]
> Also see https://github.com/love2d/love-android for some interesting stuff.

Yup, even the usual suspects like love2d, scummvm and Mupen64 are using their own implementations on top of Activity and calling native code using JNI. Probably we're the **only** app using NativeActivity.

**Comment by ildar** [comment]
@pazos , alibaba's LuaViewSDK is based on LuaJ which has a severe flaw:
https://github.com/luaj/luaj/issues/77#issuecomment-696900747
There's also `AndroLua`. It's interesting, having Java reflection API but
it's abandoned and I stuck at some serious issues like
https://github.com/jasonsantos/luajava/issues/15

**Comment by pazos** [comment]
@ildar: indeed you'll need to write JNI wrapper code anyways.

The main difference between using a self made approach and using LuaViewSDK is: with the former you'll write Activities, fragments and logic in lua and with the later you'll write everything in kotlin, define some native methods that you implement in C using JNI wrapper code and call them from kotlin/java.

In theory you should be able to put fairly large amounts of code in lua, and that code can call java/kotlin methods without any problem. It really depends on what you're trying to achieve.

Could you give us a example of logic that you want to implement in lua?. I would suggest to implement everything in kotlin unless there're very specific reasons to avoid it.

**Comment by ildar** [comment]
> indeed you'll need to write JNI wrapper code anyways.

Sorry, I don't quite catch it. Can't I just compile a module and simply
`require` it? E.g. I need luasocket first of all.

I want to avoid Kotlin/Java and even C having as much as possible in Lua.
My tasks are non-UI at the 1st stage. UI will be of interest a bit later.

**Comment by pazos** [comment]
I was talking about the ways to use lua in a android app. Of course you can still rely on this repo, but keep in mind `non-UI` things would mean a black fullscreen activity while executing lua code and you'll need to write your own `UI` primitives in LuaJIT using [`ANativeWindow` API](https://developer.android.com/ndk/reference/group/a-native-window)

When in doubt build the example with `NDK=/path/to/ndk make example` and see if it fits what you're trying to achieve.

**Comment by ildar** [comment]
Seems this discussion got some amount of off-topic. Yet TL;DR; for it:
1. to get a working apk one needs either `make example` or populate assets
with some other Lua code.
2. Actually extending this luajit Android launcher is possible with Lua
modules and even binary modules (I managed to build luasocket for it BTW)
3. There are some other projects "recommended to consider" but all have
their peculiarities. I like this one ATM.
